### PR TITLE
[Opt merge part1] Introduce columnwise merge algorithm

### PR DESCRIPTION
**Pull request by murphyatwork**
## What type of PR is this：
- [ ] bug
- [ ] feature
- [x] enhancement
- [ ] others

## Summary
Implement a column-wise merge algorithm:
1. Merge the first column, record the equal rows into a equal-range
2. Merge the second column within the equal-range of previous column
3. Repeat until no equal-range or last column

The advantage of this merge algorithm is:
1. Eliminate the `compare` function call during merge
2. Improve the memory access efficiency by sequential iterate two columns

## Experiment

TODO

### CODE REVIEW

**Comment by mofeiatwork** [comment]
@Mergify rebase

**Comment by mofeiatwork** [comment]
@Mergifyio rebase

File: be/src/exec/vectorized/sorting/merge_column.cpp [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+    Status do_visit(const ColumnType&) {
+        do_merge<ColumnType>();
+        return Status::OK();
+    }
+
+    template <class ColumnType>
+    EqualRange::Range fetch_left(const ColumnType* left_col, size_t lhs, size_t lhs_end) {
+        uint32_t first = lhs;
+        uint32_t last = lhs + 1;
+        while (last < lhs_end && left_col->compare_at(last - 1, last, *left_col, _null_first) == 0) {
```
Comment (kangkaisen) [inline be/src/exec/vectorized/sorting/merge_column.cpp:106]: If we only need to check the equal for the same column. I think we could implement a new method equal_at to replace compare_at , then there should be less branch miss.

```
struct SorterComparator {
    static int compare(const T& lhs, const T& rhs) {
        if (lhs == rhs) {
            return 0;
        } else if (lhs < rhs) {
            return -1;
        } else {
            return 1;
        }
    }
};
```

File: be/src/exec/vectorized/sorting/sort_permute.h [hunk 2]
```diff
@@ -24,6 +24,7 @@ struct PermutationItem {
     uint32_t permutation_index;
```
Comment (kangkaisen) [inline be/src/exec/vectorized/sorting/sort_permute.h:24]: Could we remove the `permutation_index` field currently?

File: be/src/exec/vectorized/sorting/sorting.h [hunk 3]
```diff
@@ -50,5 +52,92 @@ void build_tie_for_column(const ColumnPtr column, Tie* tie);
 // Append rows from permutation
 void append_by_permutation(Column* dst, const Columns& columns, const Permutation& perm);
 void append_by_permutation(Chunk* dst, const std::vector<ChunkPtr>& chunks, const Permutation& perm);
+void append_by_permutation(Chunk* dst, const std::vector<const Chunk*>& chunks, const Permutation& perm);
+
+struct SortDesc {
+    int sort_order;
```
Comment (kangkaisen) [inline be/src/exec/vectorized/sorting/sorting.h:58]: why not use bool type for sort_order and null_first?

File: be/src/exec/vectorized/sorting/sorting.h [hunk 4]
```diff
@@ -50,5 +52,92 @@ void build_tie_for_column(const ColumnPtr column, Tie* tie);
...
+    SortDesc get_column_desc(int col) const {
+        SortDesc desc;
+        DCHECK_LT(col, sort_orders.size());
+        desc.sort_order = sort_orders[col];
+        desc.null_first = null_firsts[col];
+        return desc;
+    }
+};
+
+struct SortedRun {
```
Comment (kangkaisen) [inline be/src/exec/vectorized/sorting/sorting.h:82]: would better add some comments for this struct.

File: be/src/exec/vectorized/sorting/sort_column.cpp [hunk 5]
```diff
@@ -458,6 +458,14 @@ Status sort_vertical_chunks(const bool& cancel, const std::vector<Columns>& vert
 }
 
 void append_by_permutation(Chunk* dst, const std::vector<ChunkPtr>& chunks, const Permutation& perm) {
+    std::vector<const Chunk*> src;
+    for (auto& chunk : chunks) {
```
Comment (kangkaisen) [inline be/src/exec/vectorized/sorting/sort_column.cpp:462]: how many chunks at most here?

File: be/src/exec/vectorized/sorting/sort_permute.h (hunk 2 above)
Comment (mofeiatwork) [reply be/src/exec/vectorized/sorting/sort_permute.h]: done

File: be/src/exec/vectorized/sorting/sorting.h (hunk 3 above)
Comment (mofeiatwork) [reply be/src/exec/vectorized/sorting/sorting.h:58]: actually bool type is more complicated: 

```c++
    for (size_t i = 0; i < col_num; ++i) {
        _sort_order_flag[i] = (*is_asc)[i] ? 1 : -1;
        if ((*is_asc)[i]) {
            _null_first_flag[i] = (*is_null_first)[i] ? -1 : 1;
        } else {
            _null_first_flag[i] = (*is_null_first)[i] ? 1 : -1;
        }
    }
```

File: be/src/exec/vectorized/sorting/sort_column.cpp (hunk 5 above)
Comment (mofeiatwork) [reply be/src/exec/vectorized/sorting/sort_column.cpp:462]: no limitation

File: be/src/exec/vectorized/sorting/sorting.h (hunk 4 above)
Comment (mofeiatwork) [reply be/src/exec/vectorized/sorting/sorting.h:82]: done

File: be/src/exec/vectorized/sorting/sort_column.cpp (hunk 5 above)
Comment (kangkaisen) [reply be/src/exec/vectorized/sorting/sort_column.cpp:462]: if no limit, I think we should reserve the vector.

**Comment by mofeiatwork** [comment]
@Mergifyio rebase

File: be/src/exec/vectorized/sorting/sort_column.cpp (hunk 5 above)
Comment (mofeiatwork) [reply be/src/exec/vectorized/sorting/sort_column.cpp:462]: good advice

File: be/src/exec/vectorized/sorting/merge_column.cpp (hunk 1 above)
Comment (mofeiatwork) [reply be/src/exec/vectorized/sorting/merge_column.cpp]: done. I will use the `operator ==` to replace `compare`.

File: be/src/exec/vectorized/sorting/sorting.h [hunk 6]
```diff
@@ -50,5 +52,111 @@ void build_tie_for_column(const ColumnPtr column, Tie* tie);
...
+                           size_t end);
+void append_by_permutation(Chunk* dst, const std::vector<const Chunk*>& chunks, const Permutation& perm);
+
+struct SortDesc {
+    int sort_order;
+    int null_first;
+};
+
+struct SortDescs {
+    std::vector<int> sort_orders;
```
Comment (trueeyu) [inline be/src/exec/vectorized/sorting/sorting.h:65]: why not use std::vector<SortDesc>?

File: be/src/exec/vectorized/sorting/merge_column.cpp [hunk 7]
```diff
@@ -0,0 +1,354 @@
...
+            DCHECK_EQ(rhs, rhs_end);
+        }
+
+        _equal_ranges->swap(next_ranges);
+        return Status::OK();
+    }
+
+    template <class Equal>
+    EqualRange::Range fetch_equal(size_t lhs, size_t lhs_end, Equal equal) {
+        uint32_t first = lhs;
```
Comment (trueeyu) [inline be/src/exec/vectorized/sorting/merge_column.cpp:95]: May be use exceed memory

File: be/src/exec/vectorized/sorting/merge_column.cpp [hunk 8]
```diff
@@ -0,0 +1,354 @@
...
+                        (*_perm)[output_index++] = PermutationItem(kLeftIndex, i);
+                    }
+                }
+                if (x >= 0) {
+                    rhs = right_range.second;
+                    for (size_t i = right_range.first; i < right_range.second; i++) {
+                        (*_perm)[output_index++] = PermutationItem(kRightIndex, i);
+                    }
+                }
+                if (x == 0) {
```
Comment (trueeyu) [inline be/src/exec/vectorized/sorting/merge_column.cpp:80]: This code doesn't seem to work

File: be/src/exec/vectorized/sorting/merge_column.cpp [hunk 9]
```diff
@@ -0,0 +1,354 @@
...
+                int x = 0;
+                if (lhs < lhs_end && rhs < rhs_end) {
+                    x = cmp(left_range.first, right_range.first);
+                } else if (lhs < lhs_end) {
+                    x = -1;
+                } else if (rhs < rhs_end) {
+                    x = 1;
+                }
+
+                if (x <= 0) {
```
Comment (trueeyu) [inline be/src/exec/vectorized/sorting/merge_column.cpp:68]: if x < 0, right range will fetch_equal next round?

File: be/src/exec/vectorized/sorting/merge_column.cpp [hunk 10]
```diff
@@ -0,0 +1,354 @@
...
+                    for (size_t i = 0; i < left_rows; i++) {
+                        output->emplace_back(kLeftChunkIndex, i + left_run.range.first);
+                    }
+                }
+            } else {
+                std::vector<EqualRange> equal_ranges;
+                equal_ranges.emplace_back(left_run.range, right_run.range);
+                size_t count = left_run.range.second + right_run.range.second;
+                output->resize(count);
+                equal_ranges.reserve(std::max((size_t)1, count / 4));
```
Comment (trueeyu) [inline be/src/exec/vectorized/sorting/merge_column.cpp:235]: why is count / 4 not count / 2

File: be/src/exec/vectorized/sorting/merge_column.cpp [hunk 11]
```diff
@@ -0,0 +1,354 @@
...
+    while (current.size() > 1) {
+        std::vector<ChunkPtr> next_level;
+        int level_size = current.size() & ~1;
+
+        for (int i = 0; i < level_size; i += 2) {
+            Permutation perm;
+            ChunkPtr left = current[i];
+            ChunkPtr right = current[i + 1];
+            RETURN_IF_ERROR(merge_sorted_chunks_two_way(descs, left, right, &perm));
+            ;
```
Comment (trueeyu) [inline be/src/exec/vectorized/sorting/merge_column.cpp:307]: remove `;`

File: be/src/exec/vectorized/sorting/merge_column.cpp [hunk 12]
```diff
@@ -0,0 +1,354 @@
...
+    CHECK(false) << "TODO";
+    return Status::NotSupported("TODO");
+}
+
+// Merge multiple chunks in two-way merge
+Status merge_sorted_chunks(const SortDescs& descs, const std::vector<ChunkPtr>& chunks, ChunkPtr* output) {
+    std::vector<ChunkPtr> current(chunks);
+    while (current.size() > 1) {
+        std::vector<ChunkPtr> next_level;
+        int level_size = current.size() & ~1;
```
Comment (trueeyu) [inline be/src/exec/vectorized/sorting/merge_column.cpp:300]: Will writing this improve performance?

File: be/src/exec/vectorized/sorting/merge_column.cpp (hunk 9 above)
Comment (mofeiatwork) [reply be/src/exec/vectorized/sorting/merge_column.cpp:70]: fixed

File: be/src/exec/vectorized/sorting/merge_column.cpp (hunk 12 above)
Comment (mofeiatwork) [reply be/src/exec/vectorized/sorting/merge_column.cpp:301]: not for performance, this expression is equivalent to `current.size() - current.size() % 2`, but more compact

File: be/src/exec/vectorized/sorting/merge_column.cpp (hunk 11 above)
Comment (mofeiatwork) [reply be/src/exec/vectorized/sorting/merge_column.cpp:308]: done

**Comment by mofeiatwork** [comment]
@Mergifyio rebase

**Comment by mofeiatwork** [comment]
run starrocks_clang-format

File: be/test/exec/vectorized/chunks_sorter_bench_test.cpp [hunk 13]
```diff
@@ -237,6 +269,155 @@ static void do_bench(benchmark::State& state, SortAlgorithm sorter_algo, Compare
...
+                    // state.PauseTiming();
+                    auto cloned = base_chunk->clone_unique();
+                    for (auto& col : cloned->columns()) {
+                        Int32Column::Ptr intcolumn = ColumnHelper::as_column<Int32Column>(col);
+                        int start = chunk_input_index[i] * config::vector_chunk_size;
+                        for (auto& x : intcolumn->get_data()) {
+                            x += start;
+                        }
+                    }
+                    if (output) *output = cloned.release();
```
Comment (Pslydhh) [inline be/test/exec/vectorized/chunks_sorter_bench_test.cpp:321]: Please use '{}' surround 'if' statement

File: be/src/exec/vectorized/sorting/merge_column.cpp (hunk 10 above)
Comment (mofeiatwork) [reply be/src/exec/vectorized/sorting/merge_column.cpp:236]: Which is just a empirical value.

File: be/src/exec/vectorized/sorting/sorting.h (hunk 6 above)
Comment (mofeiatwork) [reply be/src/exec/vectorized/sorting/sorting.h:65]: done

**Review by kangkaisen — APPROVED** [review]

File: be/test/exec/vectorized/chunks_sorter_bench_test.cpp (hunk 13 above)
Comment (mofeiatwork) [reply be/test/exec/vectorized/chunks_sorter_bench_test.cpp]: done

**Review by Pslydhh — APPROVED** [review]

**Review by trueeyu — APPROVED** [review]

**Pull request merged**